### ISSUE TITLE
Publish particles and state updates more frequently

### ISSUE THREAD

**Issue by guilhermelawless** [guilhermelawless/pfuclt_omni_dataset#8]
Evaluation package needs to know the current state belief even if not all observations are taken into consideration. Currently, at the beginning of a new iteration, we only move forward to fuseRobots() when everyone has predicted and given landmark measurements, and to fuseTarget() when everyone has target measurements.

This does imply that a certain hard limit will have to be set on the time between measurements given by the robot, to maintain the time sync.

Another idea is to have a fixed iteration time and if no new measurements are available at that time it will use the ones in the buffer (the most recent).

@aamirahmad what do you think? Fixed or threshold to iteration time? Or something else?

**Comment by aamirahmad** [comment]
Primarily, we cannot achieve the same synchronization as that of non-ros version simply because there we always had all the data all the time. In ros version we only have data up to a given time and not after that. So we cannot do the forward-backward sync.

Therefore, my suggestions/strategy would be as follows.
1. Robot state should be published in sync with the highest frequency sensor on it. In this case it is the odometry. Therefore, we should sync the estimates with that. 
2. Obviously the odometry timestamps of the teammates are different from the self-odometry. So in my opinion we should treat them as follows. Let me know if it makes no sense or is un-implementable.
   2.1 Use buffers for all teammate measurements and self landmark and target measurements.
   2.2 After every odometry measurement by self, run fuseRobot() and fuseTarget() with all that you have in the buffer so far.
   2.3 Publish the full-state at the timestamp same as the latest-used self-odometry.
   2.4 Obviously this means that the run-time of the iteration should be less than 1/frequency of the self-odometry. But this should be doable, or? This could be one issue in the current implementation as well because I often see the update in rviz super slow.
   2.5 Usually there should not be more than one value for any of the measurements in the buffer. If so, simply use the latest one. 
3. Finally, if pf iteration run-time is too much, we have to test by limiting the rosbag publish rate! :D

Lets close the issue only when sync issue is solved once for all!

**Comment by guilhermelawless** [comment]
1. Absolutely, estimate at the frequency of odometry, or basically each time a prediction step is called for the main robot, i'll call the estimate step.
2. Currently the PF doesn't know which one is a self robot and teammate robot, but from what you've suggested it would make sense that it knows. Will add it.
   
     2.1. Already implemented, with no difference between self and teammates, will probably add a check for the PF to know it's dealing with a self robot instead of keeping in different variables.
   
     2.2. Will add.
   
     2.3. & 2.4. What would define the iteration time, so that I can measure it? Basically only fuseRobot(), fuseTarget() and resample() ? The predict and estimate steps shouldn't count? I'll benchmark the times and see if it's doable 👍 
   
     2.5. Currently this is the behaviour.

I'll get back to you on Thursday then!

**Comment by guilhermelawless** [comment]
As of 2f83d79e798a331db69d3db1fcc4a8cd96480957 :
- The iteration time is reported with a ROS_INFO as the time it took to fuseRobots(), fuseTarget(), resample() and estimate(). Nothing else!
- Using the Eigen methods, fuseTarget() was taking way too long (almost as 4x without Eigen), so I removed that and am keeping the non-ROS method
- Iteration times are reported around 20ms (50Hz) vs. the odometry's 30ms / 30Hz. Yay! For reference, this is on a i7-4710HQ
- Currently achieving around 0.15 mean error (depending on robot) without fuseTarget(). Can't achieve the 0.106 for OMNI1 that I see in the paper, but..

@aamirahmad could you test it out and maybe figure out if the remaining issues happen because of time sync? For now with `#define DONT_FUSE_TARGET true`  in particles.cpp

**Comment by guilhermelawless** [comment]
As of e2ed40ea84c6bd1e8299329d6e9d5b9ca7beb346:
- The target fusing calculations have been checked and are ok
- The target velocity estimator has been fixed, it was previously being fed positively with the target belief instead of direct observations, resulting in disaster
- Achieving very acceptable results! However, sometimes there is quite a large offset between the particle cloud and the GT pose of the ball. Maybe there's no solution to this, it's due to the observation error. But we should check this @aamirahmad .

I will keep this open until you give the ok on the issue.

**Comment by aamirahmad** [comment]
(Sorry for the batch reply. I am back to this project only for a bit this morning and then I will look again later in the evening. Then again tomorrow during and after our meeting)

Quick comments and questions:
1. Seems like great work so far (did not yet run the code myself)!
2. Why is prediction not part of iteration time?
3. Didn't quite get the problem you meant with velocity!
4. To your last point: Perhaps I will look hands on at the code at the end of today's afternoon and then comment if there is something I figure out.

**Comment by guilhermelawless** [comment]
1. Fixed in 5a648556117d971388abc63ec26d2dd184863c8c, but only the prediction of the main robot. RIght now the predictions are made right after odometry callbacks (this has the positive result of lowering the stress during the other steps). It wouldn't be "fair" to start counting the iteration time immediately after the previous one is finished, as it will take some time for all the callbacks. So i'm counting it as:
   `iterationTime := predict(mainRobot) + predictTarget() + fuseRobots() + fuseTarget() + resample() + estimate()`
2. Velocity estimator greatly affects the performance of the target tracking. I have to discuss this with you tomorrow. Specifically: https://github.com/guilhermelawless/pfuclt_omni_dataset/commit/e2ed40ea84c6bd1e8299329d6e9d5b9ca7beb346#diff-d6b83b071f472144975555b2cdb15f93R163 . <-- **discuss tomorrow**
3. Thanks! I made a pull request for your evaluate package that now publishes the ball error as well, as well as a Matlab script to gather the performance results. Unfortunately these are not that great even though it looks good on rviz.
